src/ripple/ledger/impl/CachedView.cpp

```diff
@@ -33,25 +33,38 @@ CachedViewImpl::exists(Keylet const& k) const
 std::shared_ptr<SLE const>
 CachedViewImpl::read(Keylet const& k) const
 {
-    {
+    static CountedObjects::Counter hits{"CachedView::cache_hit"};
+    static CountedObjects::Counter hitsexpired{"CachedView::cache_hit_expired"};
+    static CountedObjects::Counter misses{"CachedView::cache_miss"};
+    bool cacheHit = false;
+    bool baseRead = false;
+
+    auto const digest = [&]() -> std::optional<uint256> {
         std::lock_guard lock(mutex_);
         auto const iter = map_.find(k.key);
         if (iter != map_.end())
         {
-            if (!iter->second || !k.check(*iter->second))
-                return nullptr;
+            cacheHit = true;
             return iter->second;
         }
-    }
-    auto const digest = base_.digest(k.key);
+        return base_.digest(k.key);
```

Comment:
If I read this right, the `digest` operation is expensive since it reads the whole of the object, and is the reason why we use `map_`. However before this change, `digest` call was outside of the scope of the critical section and now it is within this scope, making it unnecessarily longer than it needs to be. I would add an extra scope right before `lock_guard` and close it just above this line (just like it was done before this change).

Alternatively we could just extend this critical section to the whole of the function, as it will have some benefits (just one opportunity to block on mutex, instead of two), but it is very much a speculation whether or not that would be an improvement.

Reply:
Good catch! Yes, I'd rather keep the critical section as narrow as possible, if only to prevent potential problems down the road.